Cross-validation option added

ArantxaCasanova opened the pull request:
To be able to separate de data in folds and select a specific one for validation.

ArantxaCasanova commented:
I already applied the coding style modifications.
If there is anything else, don't hesitate to ask. 
Thank you!

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -33,8 +33,14 @@ class IsbiEmStacksDataset(ThreadedDataset):
-        whereas 15\% will be used for validation.
+        For example, if split=0.85, 85\% of the images will be used
+        for training, whereas 15\% will be used for validation.
+    crossval: int
```
int or None

fvisin replied:
Also please rename as `crossval_nfolds`

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -33,8 +33,14 @@ class IsbiEmStacksDataset(ThreadedDataset):
+        for training, whereas 15\% will be used for validation.
+    crossval: int
+         If it is set to None, to cross-validation is used. An int specifying
+         in how many folds we want to split our data.
```
When None cross-validation is disabled. Else, represents the number of folds we data will be split into.

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
     _mask_labels = {0: 'Non-membranes', 1: 'Membranes'}
 
-    def __init__(self, which_set='train', split=0.85, *args, **kwargs):
+    def __init__(self, which_set='train', split=0.60, crossval=5, fold=3, rand_perm=None,
```
The description of `rand_perm` is missing.

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -33,8 +33,14 @@ class IsbiEmStacksDataset(ThreadedDataset):
+         in how many folds we want to split our data.
+    fold: int
+        An int specifying which fold we want. If fold=1, images from 0 to 5
+        will be used as validation. If fold=2, images from 6 to 11, and so on.
```
`fold` should be zero-based. Please change it to:
> An int specifying which fold to use for validation. If fold=0, images from 0 to 5 will be used for validation. If fold=1, images from 6 to 11, and so on.

fvisin replied:
Rename to `valid_fold`

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
-        elif self.which_set == "test":
-            self.start = 0
-            self.end = 30
+        self.middle_fold = False  # False by default
```
Please remove the comment, the code is already clear.

fvisin replied:
Please remove self.middle_fold altogether (see comment below)

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
-            self.start = 0
-            self.end = 30
+        self.middle_fold = False  # False by default
+        if crossval is not None:  # if cross-validation is used
```
Please remove this comment as well, the code is already clear.

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+        self.middle_fold = False  # False by default
+        if crossval is not None:  # if cross-validation is used
+            if rand_perm is not None:
+                self.rand_indexes=rand_perm
```
Please rename to `rand_indices`

fvisin replied:
Also add a space around the `=` operator

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+            else:
+                self.rand_indexes=range(0,30)
+            # number of images per fold
+            img_per_fold = int(30/crossval)
```
Please add a space before and after the `/` operator

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+            # number of images per fold
+            img_per_fold = int(30/crossval)
+            # start and end index for validation fold
+            start = (fold-1)*img_per_fold
```
Please add a space around the `*` operator and remove the -1 (we start from 0 now)
Edit: disregard, see comment below.

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+            img_per_fold = int(30/crossval)
+            # start and end index for validation fold
+            start = (fold-1)*img_per_fold
+            end = fold*img_per_fold
```
make it `(fold + 1) * img_per_fold`
Edit: disregard, see comment below.

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -33,8 +33,14 @@ class IsbiEmStacksDataset(ThreadedDataset):
-        For example, if split=0.85, 85\% of the images will be used for training,
-        whereas 15\% will be used for validation.
+        For example, if split=0.85, 85\% of the images will be used
+        for training, whereas 15\% will be used for validation.
```
Add "Will be ignored if crossval_nfolds is not None"

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+            elif self.which_set == "val":
+                self.start = start
+                self.end = end
+            elif self.which_set == "test":
```
I don't think 'test' makes sense in the cross-validation case. I suggest to raise a ValueError if which_set is 'test' here.

fvisin replied:
More generally, I suggest to replace L75-93 with the following approach:
```python
if self.which_set == "train":
    self.start_1 = 0
    self.end_1 = fold * img_per_fold
    self.start_2 = (fold+1) * img_per_fold
    self.end_2 = 30
elif self.which_set == "val":
    self.start_1 = fold * img_per_fold
    self.end_1 = self.start_2 = self.end_2 = (fold+1) * img_per_fold
elif self.which_set == "test":
    raise ValueError('Cannot perform cross-validation on test.')
```
and then replace self.get_names with
```python
return {'default': self.rand_indexes[self.start_1:self.end_1] + 
                   self.rand_indexes[self.start_2:self.end_2]}
```

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -52,20 +58,52 @@ class IsbiEmStacksDataset(ThreadedDataset):
+            if rand_perm is not None:
+                self.rand_indexes=rand_perm
+            else:
+                self.rand_indexes=range(0,30)
```
self.rand_indexes = range(30)

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -81,7 +119,12 @@ def __init__(self, which_set='train', split=0.85, *args, **kwargs):
+            # if validation is a middle fold, concatenate separated train folds
+            return {'default': self.rand_indexes[range(0, self.start)+range(self.end, 30)].tolist()}
+        else:
+            return {'default': self.rand_indexes[range(self.start, self.end)].tolist()}
```
Replace with the suggested command (see comment above)

pep8speaks commented:
Hello @ArantxaCasanova! Thanks for updating the PR.



Cheers ! There are no PEP8 issues in this Pull Request. :beers: 

##### Comment last updated on July 05, 2017 at 16:58 Hours UTC

ArantxaCasanova commented:
Thanks for the revision! I corrected what you suggested.

I still had to use "self.crossval = True" to control the "get_names()" function without making it too messy.
I also added the description for the "rand_perm" variable, I forgot to add it before.

Let me know what you think!

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -120,12 +118,13 @@ def __init__(self, which_set='train', split=0.60, crossval=5, fold=3, rand_perm=
+            return {'default': (
+                                  self.rand_indices[self.start_1:self.end_1]
+                              ).tolist() + (
+                self.rand_indices[self.start_2:self.end_2]).tolist()}
```
The indentation makes it difficult to read the code.
This should respect the PEP8 while maintaining the readability of the code:
```python
return {'default':
                      self.rand_indices[self.start_1:self.end_1]).tolist() + (
                      self.rand_indices[self.start_2:self.end_2]).tolist()}
```

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -59,43 +65,35 @@ class IsbiEmStacksDataset(ThreadedDataset):
-        self.middle_fold = False  # False by default
-        if crossval is not None:  # if cross-validation is used
+        if crossval_nfolds is not None:
+            self.crossval = True
```
I would prefer not to add this attribute to the class since the dataset objects are already a bit cluttered :)
I suggest to remove it and check for `hasattr(self, 'rand_indices')` in `get_names()`

fvisin commented on dataset_loaders/images/isbi_em_stacks.py:
```diff
@@ -120,12 +118,13 @@ def __init__(self, which_set='train', split=0.60, crossval=5, fold=3, rand_perm=
-        if self.middle_fold:
-            # if validation is a middle fold, concatenate separated train folds
-            return {'default': self.rand_indexes[range(0, self.start)+range(self.end, 30)].tolist()}
+        if self.crossval:
```
See comment above
